### PR TITLE
Authenticate with es

### PR DESCRIPTION
**Pull request by rock-yu**
## Proposed changes

Add command-line options to provide credentials when using secured ES 

[#38  Provide options to specify ES basic authentications to verify against secured ES Cluster](https://github.com/Aconex/scrutineer/issues/38)

## Checklist

- [x] I've read the [CONTRIBUTING](https://github.com/Aconex/scrutineer/blob/master/CONTRIBUTING.md) doc
- [x] All tests pass locally with my changes
- [x] I've added unit tests and integration tests.
- [ ] I've added necessary documentation (if appropriate)

### CODE REVIEW

File: src/main/java/com/aconex/scrutineer/CmdLineOptionsSSLVerificationModeValidator.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import com.beust.jcommander.IValueValidator;
+import com.beust.jcommander.ParameterException;
+
+public class CmdLineOptionsSSLVerificationModeValidator implements IValueValidator {
+    private final String[] sslVerificationModes = {"none", "certificate", "full"};
+
+    @Override
+    public void validate(String name, Object value) throws ParameterException {
+        if (!Arrays.asList(sslVerificationModes).contains(value)) {
+            throw new ParameterException("SSL verification mode can be one of: " + Arrays.toString(sslVerificationModes));
```
Comment (tallpsmith) [inline src/main/java/com/aconex/scrutineer/CmdLineOptionsSSLVerificationModeValidator.java:14]: always good to pass back what was received that is invalid. something like:
```
..."Invalid SSL Verification mode '%s' - can be one of: %s", value, Arrays.toString(sslVerificationModes) ...
```

File: src/main/java/com/aconex/scrutineer/ScrutineerCommandLineOptions.java [hunk 2]
```diff
@@ -16,6 +16,19 @@
...
     public List<TransportAddress> elasticSearchHosts;
 
+    @Parameter(names = "--esUsername", description = "Elasticsearch Username")
+    public String esUsername;
+
+    @Parameter(names = "--esPassword", description = "Elasticsearch Password", password = true)
+    public String esPassword;
+
+    @Parameter(names = "--esSSLVerificationMode",
+            description = "Transport client SSL verification mode, accepted values: [none, certificate, full]", validateValueWith = CmdLineOptionsSSLVerificationModeValidator.class)
```
Comment (tallpsmith) [inline src/main/java/com/aconex/scrutineer/ScrutineerCommandLineOptions.java:26]: Do we need to outline in the `description` that the default value is `certificate` ?

Comment (tallpsmith) [reply src/main/java/com/aconex/scrutineer/ScrutineerCommandLineOptions.java:26]: or does the JCommander stuff do that for us (I forgot! )

File: src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                    .put("xpack.security.transport.ssl.verification_mode", sslVerificationMode);
+        }
+
+        return settings
+                .put(ClusterName.CLUSTER_NAME_SETTING.getKey(), clusterName)
+                .put("transport.tcp.connect_timeout", SOCKET_CONNECT_TIMEOUT);
+    }
+
+    private void addTransportHosts(TransportClient transportClient, List<TransportAddress> elasticSearchHosts) {
+        LogUtils.debug(LOG, "Starting elastic search on hosts: %s ", elasticSearchHosts.toString());
```
Comment (tallpsmith) [inline src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java:82]: This log message seems odd.. It's not really "Starting" ES, but more "Connecting" to ES on these hosts isn't it?

File: src/test/java/com/aconex/scrutineer/CompositeStreamVerifierListenerTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    private CompositeStreamVerifierListener testInstance;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+        this.testInstance = new CompositeStreamVerifierListener(Arrays.asList(otherListener1, otherListener2));
+    }
+
+    @Test
+    public void testShouldDelegateToAllListenersOnPrimaryStreamProcessed() {
```
Comment (tallpsmith) [inline src/test/java/com/aconex/scrutineer/CompositeStreamVerifierListenerTest.java:31]: This is more a style comment, but given each test method is annotated with the `@Test` marker, I prefer to remove the prefix `test` from each method for readability.    I think it reads nicer.

**Review by tallpsmith — CHANGES_REQUESTED** [review]
Looks pretty good!

File: src/main/java/com/aconex/scrutineer/CmdLineOptionsSSLVerificationModeValidator.java (hunk 1 above)
Comment (rock-yu) [reply src/main/java/com/aconex/scrutineer/CmdLineOptionsSSLVerificationModeValidator.java]: 👍 Fixed in code

File: src/main/java/com/aconex/scrutineer/ScrutineerCommandLineOptions.java (hunk 2 above)
Comment (rock-yu) [reply src/main/java/com/aconex/scrutineer/ScrutineerCommandLineOptions.java:29]: I've added new Jcommander parameter "--help or -h" to print out the usage. 

Yes, JCommander will printout which is the default:



```
 ./target/appassembler/bin/scrutineer -h
Usage: <main class> [options]
  Options:
  * --clusterName
      ElasticSearch cluster name identifier
  * --esHosts
      CSV set of seed ElasticSearch host:port pairs to use as part of 
      discovery 
    --esPassword
      Elasticsearch Password
    --esSSLEnabled
      Enable SSL encryption for transport client
      Default: false
    --esSSLVerificationMode
      Transport client SSL verification mode, accepted values: [none, 
      certificate, full]
      Default: certificate

```

File: src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java (hunk 3 above)
Comment (rock-yu) [reply src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java]: Good point, fixed in code

File: src/test/java/com/aconex/scrutineer/CompositeStreamVerifierListenerTest.java (hunk 4 above)
Comment (rock-yu) [reply src/test/java/com/aconex/scrutineer/CompositeStreamVerifierListenerTest.java]: It was a mistake to add `test` prefix. Fixed in code

File: src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java [hunk 5]
```diff
@@ -79,7 +79,7 @@ private Settings createSettings(Credential esCredentials, String sslVerification
     }
 
     private void addTransportHosts(TransportClient transportClient, List<TransportAddress> elasticSearchHosts) {
-        LogUtils.debug(LOG, "Starting elastic search on hosts: %s ", elasticSearchHosts.toString());
+        LogUtils.debug(LOG, "Connecting to elastic search on hosts: %s ", elasticSearchHosts.toString());
```
Comment (tallpsmith) [inline src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java:82]: `pedanticMode: On`

Can you make that string `Elasticsearch`... ? :-D

**Review by tallpsmith — CHANGES_REQUESTED** [review]
Minor string fixup and we're done!

File: src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java (hunk 5 above)
Comment (rock-yu) [reply src/main/java/com/aconex/scrutineer/elasticsearch/ElasticSearchTransportClientFactory.java]: Sure, I've renamed "elastic search" to "Elasticsearch"

**Review by tallpsmith — APPROVED** [review]

**Pull request merged**